src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractGenerateNuGetPackageCommand.cs

```diff
@@ -1,40 +1,151 @@
+        protected AbstractGenerateNuGetPackageCommand(
+            UnconfiguredProject unconfiguredProject,
+            IProjectThreadingService threadingService,
+            SVsServiceProvider serviceProvider,
+            GeneratePackageOnBuildPropertyProvider generatePackageOnBuildPropertyProvider)
         {
+            Requires.NotNull(unconfiguredProject, nameof(unconfiguredProject));
+            Requires.NotNull(threadingService, nameof(threadingService));
+            Requires.NotNull(serviceProvider, nameof(serviceProvider));
+            Requires.NotNull(generatePackageOnBuildPropertyProvider, nameof(generatePackageOnBuildPropertyProvider));
+
             UnconfiguredProject = unconfiguredProject;
             _threadingService = threadingService;
+            _serviceProvider = serviceProvider;
+            _generatePackageOnBuildPropertyProvider = generatePackageOnBuildPropertyProvider;            
         }
 
         protected UnconfiguredProject UnconfiguredProject { get; }
 
         protected abstract string GetCommandText();
         protected abstract bool ShouldHandle(IProjectTree node);
 
-        protected override Task<CommandStatusResult> GetCommandStatusAsync(IProjectTree node, bool focused, string commandText, CommandStatus progressiveStatus) =>
-            ShouldHandle(node) ?
-                GetCommandStatusResult.Handled(GetCommandText(), CommandStatus.Enabled) :
-                GetCommandStatusResult.Unhandled;
+        protected async override Task<CommandStatusResult> GetCommandStatusAsync(IProjectTree node, bool focused, string commandText, CommandStatus progressiveStatus)
+        {
+            if (ShouldHandle(node))
+            {
+                // Enable the command if the build manager is ready to build.
+                var commandStatus = await IsReadyToBuildAsync().ConfigureAwait(false) ? CommandStatus.Enabled : CommandStatus.Supported;
+                return await GetCommandStatusResult.Handled(GetCommandText(), commandStatus).ConfigureAwait(false);
+            }
+
+            return CommandStatusResult.Unhandled;
+        }
+
+        private async Task<bool> IsReadyToBuildAsync()
+        {
+            // Ensure build manager is initialized.
+            await EnsureBuildManagerInitializedAsync().ConfigureAwait(false);
+
+            ErrorHandler.ThrowOnFailure(_buildManager.QueryBuildManagerBusy(out int busy));
+            return busy == 0;
+        }
+
+        private async Task EnsureBuildManagerInitializedAsync()
+        {
+            if (_buildManager == null)
+            {
+                // Switch to UI thread for querying the build manager service.
+                await _threadingService.SwitchToUIThread();
+
+                var buildManager = _serviceProvider.GetService<IVsSolutionBuildManager2, SVsSolutionBuildManager>();
+                if (Interlocked.CompareExchange(ref _buildManager, buildManager, null) == null)
```

Comment:
As long as you always do this on the UI thread I don't think we need to use Interlocked and introduce a memory barrier, right? You just need to check if _buildManager is null again after switching to the UI thread. Although this works as well.